### ISSUE TITLE
Connection pooling with `resource-pool` and `spockT`

### ISSUE THREAD

**Issue by lunaris** [agrafix/Spock#77]
I'm using the `spockT` function to support a custom monad transformer stack but am having some issues with connection pooling as provided by the `resource-pool` package. Specifically, I have a `PostgreSQLT` transformer which wraps `postgresql-simple` and some functions for interacting with it:

```haskell
createPostgreSQLConnectionPool :: ... -> IO (Pool Connection)
runPostgreSQLT :: Pool Connection -> PostgreSQLT m a -> m a
```

The `main` function for my API looks something like:

```haskell
main 
  = do
      ...
      pool <- createPostgreSQLConnectionPool ...
      ...
      runSpock config $ spockT (runPostgreSQLT pool) $ do
        -- Define the Spock application, etc.
        ...
```

While this appears to work fine in terms of connections being created, exceptions being recognised and triggering appropriate destruction, the "reaper thread" that `resource-pool` provides is not working. After doing some digging, I can see the thread starting but never running past its initial `threadDelay`. I believe that after `spockT` and `runSpock` get going, there's no yield or opportunity for the thread to do its thing again. 

Is there any way I can remedy this? Perhaps you're aware of this and this is why Spock provides support for pooling and the like "natively" -- because bolting it on like this is difficult/unachievable? I'd be keen to hear any options, since I'd rather not have the endpoint definitions manage connection acquisition/release themselves when the transformer can capture this behaviour automatically. Thanks in advance for any feedback.

**Comment by agrafix** [comment]
From what you are writing everything looks good to me - but it does sound like you forgot to compile (and run) with the threaded RTS? Spock only builds up an `wai Application`, so as long as you are not doing anything blocking while defining your routes there should not be problems...

**Comment by lunaris** [comment]
I don't think that's what happening, but I'm far from an expert in these matters. I'm building with Stack and the `.cabal` file for the binary I'm building contains:

```
  Ghc-Options:        -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T -Wall -Werror

```

Which, as far as I understand it, means that the binary should be both compiled with the threaded RTS and run with it (implicitly passing `+RTS -N -RTS`, right?). That said, I've tried passing the RTS options myself with:

```
% stack exec my-api -- +RTS -N -RTS
```

and this doesn't appear to have made a difference. Any other thoughts/code examples I might crib? I should have made it clear in the original issue, but I'm working with `0.10.0.1` by the way. Thanks again.

**Comment by agrafix** [comment]
That looks good. If you could provide a minimum working code sample (small stack project) with the issue I'm happy to take a look at it.

**Comment by lunaris** [comment]
Well, this is an interesting one. I set about producing a minimum code sample and came up with https://github.com/lunaris/spock-resource-pool. However, this works! So I set about taking the codebase I had and paring it down to match the example and eventually found that it has nothing to do with Spock (at least, not directly). In short, my application:

* Reads some options in using `lookupEnv` (`System.Environment`) and `optparse-applicative`.
* Converts those options into a configuration which creates the pool.
* Sets up Spock as in the minimum (working) example.

As best as I can tell, GHC's inliner (!) is causing trouble in the first step. Since reading the options might fail (or options might not have been provided), I have a function `fromOptions` which returns a `Maybe (IO Configuration)` (where `Configuration` contains my connection pool). If I instrument `fromOptions` with a `putStrLn` call (or explicitly add a `{-# NOINLINE ... #-}` pragma), the pool comes out just fine and I can observe the reaper thread executing. If I don't, the pool comes out fine and is able to supply connections, but the reaper thread doesn't work (though it *is* started -- it simply never comes back after its initial `threadDelay` call). Baffling stuff.

I've put the core dumps of the different versions in the linked repository (under `dumps`) but I appreciate this is already a million miles away from being your problem (and indeed, the `NOINLINE` gets me to a working solution at least). But, if your core-fu is better than mine, or you know someone who might be able and willing to explain what's happening to my code, I would be incredibly grateful (there's no cheeky `unsafePerformIO` or anything like that I can see). All that said, I guess this issue can be closed.

Thanks for your time and the incredibly prompt responses!
